dot/sync/chain_sync.go

```diff
@@ -609,19 +678,41 @@ func (cs *chainSync) doSync(req *network.BlockRequestMessage) *workerError {
 	// response was validated! place into ready block queue
 	for _, bd := range resp.BlockData {
 		// block is ready to be processed!
-		logger.Trace("new ready block", "hash", bd.Hash, "number", bd.Header.Number)
-		cs.pendingBlocks.removeBlock(bd.Hash)
-		cs.readyBlocks <- bd
+		handleReadyBlock(bd, cs.pendingBlocks, cs.readyBlocks)
 	}
 
 	return nil
 }
 
+func handleReadyBlock(bd *types.BlockData, pendingBlocks DisjointBlockSet, readyBlocks *blockQueue) {
+	// see if there are any descendents in the pending queue that are now ready to be processed,
+	// as we have just become aware of their parent block
+
+	// if header was not requested, get it from the pending set
+	// if we're expecting headers, validate should ensure we have a header
+	if bd.Header == nil {
+		block := pendingBlocks.getBlock(bd.Hash)
+		if block.header == nil {
+			panic("block isn't ready! header is unknown!")
+		}
+
+		bd.Header = block.header
+	}
+
+	logger.Trace("new ready block", "hash", bd.Hash, "number", bd.Header.Number)
+
+	ready := []*types.BlockData{bd}
+	ready = pendingBlocks.getReadyDescendants(bd.Hash, ready)
+
+	for _, rb := range ready {
+		pendingBlocks.removeBlock(rb.Hash)
+		readyBlocks.push(rb)
```

Comment:
this call could be blocking right? Do we want to block `dispatchWorker` from progressing?

Reply:
potentially, i think it's goo if it blocks, otherwise it'll return to the result handler which may be doing some logic to check what's in the `readyBlocks` queue, so if this is still waiting then the result handler might be incorrect